src/components/Carousel.js

```diff
@@ -402,6 +430,16 @@ class Carousel extends Component {
             return currentPosition;
         }
 
+        if (this.props.infiniteLoop) {
+            if (this.state.loop && index === 0) {
+                return 0;
+            }
+            if ((this.state.loop || swiping) && index === this.childrenLength - 1) {
+                return - (this.childrenLength + 1) * 100;
+            }
+            return - (index + 1) * 100;
+        }
+
```

Comment:
I can understand this now, it is adjusting the output to "- (index + 1) * 100" which makes sense as we've inserted an extra clone at the start.

However, the this.state.loop and swiping contexts are only ever used by specific callers. For example, swiping parameter is only used during onSwipeMove so likely that logic should sit inside onSwipeMove. Then the this.state.loop seems odd as it looks like it's only used inside moveTo in order to ensure that the next render changes the behaviour of getPosition. I'm not keen on this as it means getPosition is not predictable - depending when you call it, you get different results - so it seems this.state.loop is a bit of a hack to get moveTo to behave correctly after moving to a cloned slide.

Still trying to get my head around the math here.

Reply:
Ah OK I think I get it. When sliding from 0 to -1 (in other words, one behind the first item) instead of sliding item 0 into the cloned slide, we need to jump to the final cloned slide which for 2 slide carousel is slide index 2, then slide that through to index 1 (which is same as the cloned slide at -1).

I'm not sure why we're using all these states to get this - can we not just call setPosition() to push to slide 2 during the moveTo call (escape from React is fine - we do the same during swiping) - then call setState/selectItem to trigger the change and it should then render transition and move correctly. Sounds like it would be far far more simpler and have no need to play around with this.state.loop or animation frames. Did you try this? Might be worth doing as it'll greatly simplify the logic if it works.